### PR TITLE
*: move tso to independent thread

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve?

Issue Number: Ref #8477.

### What is changed and how does it work?

This is a part of #8517.  

```commit-message
```

### Check List

Tests

- Unit test

### Release note

```release-note
None.
```

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -203,7 +203,9 @@ func (k *serviceModeKeeper) close() {
 	defer k.Unlock()
 	switch k.serviceMode {
 	case pdpb.ServiceMode_API_SVC_MODE:
-		k.tsoSvcDiscovery.Close()
+		if k.tsoSvcDiscovery != nil {
```
Comment (JmPotato) [inline client/client.go:206]: What about moving this check into `Close()`?

File: server/cluster/cluster.go [hunk 2]
```diff
@@ -390,25 +400,85 @@ func (c *RaftCluster) checkServices() {
 	}
 }
 
+// checkTSOService checks the TSO service.
+func (c *RaftCluster) checkTSOService() {
+	if !c.isAPIServiceMode {
```
Comment (JmPotato) [inline server/cluster/cluster.go:405]:
```suggestion
	if c.isAPIServiceMode {
		return
	}
```

File: server/cluster/cluster.go [hunk 3]
```diff
@@ -390,25 +400,85 @@ func (c *RaftCluster) checkServices() {
...
+		if c.member.IsLeader() {
+			if err := c.startTSOJobs(); err != nil {
+				// If there is an error, need to wait for the next check.
```
Comment (JmPotato) [inline server/cluster/cluster.go:408]: Do we need log here?

File: server/cluster/cluster.go [hunk 4]
```diff
@@ -390,25 +400,85 @@ func (c *RaftCluster) checkServices() {
...
+				return
+			}
+		} else {
+			// leader exits, reset the allocator group
+			if err := c.stopTSOJobs(); err != nil {
+				// If there is an error, need to wait for the next check.
```
Comment (JmPotato) [inline server/cluster/cluster.go:414]: Ditto.

File: server/cluster/cluster.go [hunk 5]
```diff
@@ -314,11 +322,13 @@ func (c *RaftCluster) Start(s Server) error {
 	if err != nil {
 		return err
 	}
+	c.checkTSOService()
```
Comment (okJiang) [inline server/cluster/cluster.go:325]: Can we put this func next to `checkSchedulingService`?

**Review by okJiang — COMMENTED** [review]
rest lgtm

File: server/server.go [hunk 6]
```diff
@@ -1715,29 +1714,6 @@ func (s *Server) campaignLeader() {
 	s.member.KeepLeader(ctx)
 	log.Info(fmt.Sprintf("campaign %s leader ok", s.mode), zap.String("campaign-leader-name", s.Name()))
 
-	if !s.IsAPIServiceMode() {
```
[the remark this answers is not in the document]
Comment (rleungx) [reply server/server.go:1718]: it still works

File: server/cluster/cluster.go (hunk 5 above)
Comment (rleungx) [reply server/cluster/cluster.go:325]: prefer to check it immediately

**Review by nolouch — APPROVED** [review]

File: server/cluster/cluster.go [hunk 7]
```diff
@@ -390,25 +400,76 @@ func (c *RaftCluster) checkServices() {
 	}
 }
 
+// checkTSOService checks the TSO service.
+func (c *RaftCluster) checkTSOService() {
```
Comment (okJiang) [inline server/cluster/cluster.go:404]: Seems we do not need this wrap function now. We can call `startTSOJobs` directly.

Comment (rleungx) [reply server/cluster/cluster.go:404]: for the upcomming PR

**Review by okJiang — APPROVED** [review]

**Pull request merged**